src/v/cluster/archival_metadata_stm.cc

```diff
@@ -620,9 +620,145 @@ ss::future<std::error_code> archival_metadata_stm::process_anomalies(
     co_return co_await builder.replicate();
 }
 
+ss::future<bool>
+archival_metadata_stm::sync(model::timeout_clock::duration timeout) {
+    /*
+     * Implementation note
+     * ===================
+     *
+     * This function needs to handle two cases:
+     * Case 1: we have not synced since the last term (e.g. on leadership
```

Comment:
Should this always be true because we're calling `sync` after the replica became a leader? And `sync` guarantees that all batches from all previous terms are applied IIRC.
cc @rystsov

Reply:
Right, that's what `persisted_stm::sync` does. The issue we are trying to prevent is related to stopping and starting the archiver in the same term.

Right now, we first signal the abort source and then wait for the gate to close. The following sequence of operations is possible:
1. archiver replicates some commands
2. replicate succeeds
3. we start waiting on apply
4. abort is requested by `ntp_archiver::stop`
5. waiting is aborted
6. we ended up in a case 2

An alternative solution is to make sure that `ntp_archiver::stop` leaves the STM in sync.
I think that's kind of brittle since someone can add a new path for replicating commands in the future
and not be aware of this requirement. It also relies very heavily on the correctness of replication semantics.
Hence why I picked the more generic approach of caching the latest replicated offset and waiting on it.

Reply:
This looks like a solution to a problem that shouldn't exist. Why isn't it possible to avoid stopping the archiver?